src/lexer.rs

```diff
@@ -0,0 +1,319 @@
+        )
+    }
+
+    /// Consume a component.
+    ///
+    /// A component can either be an alphanumeric or numeric.
+    /// Does not permit leading zeroes if numeric.
+    fn component(&mut self, start: usize) -> Result<Token<'input>, Error> {
+        let end = scan_while!(self, start, '0'...'9' | 'A'...'Z' | 'a'...'z');
+        let input = &self.input[start..end];
+
+        let mut it = input.chars();
+        let (a, b) = (it.next(), it.next());
+
+        // exactly zero
+        if a == Some('0') && b.is_none() {
+            return Ok(Numeric(0));
+        }
+
+        if a != Some('0') {
+            if let Ok(numeric) = input.parse::<u64>() {
+                return Ok(Numeric(numeric));
+            }
+        }
+
+        Ok(AlphaNumeric(input))
+    }
+
+    /// Consume whitespace.
+    fn whitespace(&mut self, start: usize) -> Result<Token<'input>, Error> {
+        let end = scan_while!(self, start, ' ' | '\t' | '\n' | '\r');
+        Ok(Whitespace(start, end))
+    }
+}
+
+impl<'input> Iterator for Lexer<'input> {
+    type Item = Result<Token<'input>, Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            // two subsequent char tokens.
+            if let Some((_, a, b)) = self.two() {
+                let two = match (a, b) {
+                    ('<', '=') => Some(LtEq),
+                    ('>', '=') => Some(GtEq),
+                    ('|', '|') => Some(Or),
+                    _ => None,
+                };
+
+                if let Some(two) = two {
+                    self.step_n(2);
+                    return Some(Ok(two));
+                }
+            }
+
+            // single char and start of numeric tokens.
+            if let Some((start, c)) = self.one() {
+                let tok = match c {
+                    ' ' | '\t' | '\n' | '\r' => {
+                        self.step();
+                        return Some(self.whitespace(start));
+                    }
+                    '=' => Eq,
+                    '>' => Gt,
+                    '<' => Lt,
+                    '^' => Caret,
+                    '~' => Tilde,
+                    '*' => Star,
+                    '.' => Dot,
+                    ',' => Comma,
+                    '-' => Hyphen,
+                    '+' => Plus,
+                    '0'...'9' | 'a'...'z' | 'A'...'Z' => {
+                        self.step();
+                        return Some(self.component(start));
+                    }
+                    c => return Some(Err(UnexpectedChar(c))),
+                };
+
+                self.step();
+                return Some(Ok(tok));
+            };
+
+            return None;
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn lex(input: &str) -> Vec<Token> {
+        Lexer::new(input).map(Result::unwrap).collect::<Vec<_>>()
+    }
+
+    #[test]
+    pub fn all_tokens() {
+        assert_eq!(
+            lex("=><.^~*0.1234<=>=||"),
```

Comment:
this only seems to have 13 out of the 16 tokens; whitespace is missing

also i wonder if we shouldn't check x/X in this test too

Reply:
Added coverage for all tokens, and split out tests for components (alphanumeric, numeric).
Added a separate test for testing just the `is_wildcard` method.